[#39450] Custom Field User Not available in Filter

fraber opened the pull request:
https://community.openproject.org/work_packages/39450

I've re-enabled custom fields of type "User" to act as filters in global work_packages 
report:

Custom fields of type "User" and "Version" were excluded in 2018 by Jens from
appearing in the global work_packages reports. I guess that's because their values
are set per-project.
However, it seems that the current "principal" endpoint now returns the global list of
users in this case, so that a filter of type "User" works correctly, at least on my dev
installation.

fraber committed:
Allow "User" custom fields to act as filters in global work_packages

Custom fields of type "User" and "Version" were excluded in 2018 from
appearing in the global work_packages reports. I guess that's because
their values are set per-project. However, it seems that the current
"principal" endpoint returns the global list of users now, so that a
filter of type "User" works correctly, at least in my installation.

fraber commented:
I've checked with a simple user, and this user also gets the list of all users in the system as options for the filter value.

ulferts reviewed (changes requested):
Should be working @fraber. I am a bit hesitant when it comes to the implication of potentially loading 1000 users through this field but that is something we have the tools to handle later on by now.

What is missing is a feature spec. Add that and I will merge the PR.

fraber committed:
Added a spec to check that user custom fields appear on global WP page...

... if they are defined with "For all projects" (is_for_all = true).

fraber commented:
Hi @ulferts, added the feature spec. I hope I wasn't too verbose with the comments...

ulferts reviewed (changes requested):
Since I reviewed #10552 before, I checked for the same scenario and this feature now suffers from the same bug. In this case, the change needs to be made in https://github.com/opf/openproject/blob/dev/app/models/custom_field.rb#L295 where again the `.visible` scope needs to be used.

fraber committed:
Fixed issue with visibility of users who have been invited but not
yet logged in as explained by @ulferts in:
https://github.com/opf/openproject/pull/10450#pullrequestreview-954959121

fraber commented:
Hi @ulferts, I've included the new scope for user-ranging custom fields as you proposed.
I was able to reproduce the issue as described here:
https://community.openproject.org/projects/openproject/work_packages/41934/activity
I confirm that this error message has gone after changing the scope to .visible.

fraber commented:
Writing a spec now to test for regression.

fraber committed:
Added a spec to test for changed scope of possible values of a user custom field.
I've extended the existing model spec in the style of the existing model spec...
